meteor/sofie-server/model/entities/segment.ts

```diff
@@ -91,11 +90,43 @@ export class Segment {
 		return this.parts
 	}
 
-	public addInfinitePieces(infinitePieces: Piece[]): void {
-		infinitePieces.forEach((piece: Piece) => this.infinitePieces.set(piece.layer, piece))
+	public getFirstSpanningPieceForEachLayerBeforePart(part: Part, layersToIgnore: Set<string>): Piece[] {
+		const indexOfPart = this.parts.findIndex((p) => p.id === part.id)
+		return this.getPiecesFromIndex(indexOfPart - 1, layersToIgnore, [
+			PieceLifespan.SPANNING_UNTIL_RUNDOWN_END,
+			PieceLifespan.SPANNING_UNTIL_SEGMENT_END,
+		])
+	}
+
+	private getPiecesFromIndex(
+		startIndex: number,
+		layersToIgnore: Set<string>,
+		lifespanFilters: PieceLifespan[]
+	): Piece[] {
+		const pieces: Piece[] = []
+
+		for (let i = startIndex; i >= 0; i--) {
+			const piecesSpanningSegment: Piece[] = this.parts[i].getPiecesWithLifespan(lifespanFilters)
+			for (let j = piecesSpanningSegment.length - 1; j >= 0; j--) {
+				const piece: Piece = piecesSpanningSegment[j]
+				if (layersToIgnore.has(piece.layer)) {
+					continue
+				}
+				pieces.push(piece)
+				layersToIgnore.add(piece.layer)
+			}
+		}
+
+		return pieces
+	}
```

Comment:
This seems quite complex and I am not sure what it should do based on the name. Can we have a chat about this?

Reply:
I agree, the name doesn't do the function justice.

In summary:
It finds all Pieces on the Segment that is _not_ on a layer that's in `layersToIgnore` and has a Lifespan that is in `lifespansFilters`. 
Whenever a Piece is found, the layer of the Piece is added to `layersToIgnore` because we only want to find one Piece per layer.
It starts looking in the Parts array at the index specified by `startIndex` and works it's way towards the beginning of the array.

The usecase is when we have jumped from the 1st to (for example) the 4th Part in a Segment, we need find infinite Pieces that is not on any layers from the 4th Part but are in Part 3 or 2 etc.

Suggestions are most welcome :)

Reply:
I have taken the liberty to come with a suggestion that uses a more functional approach, where `getPiecesFromIndex` is renamed to `getPiecesOnUnusedLayersFromIndexToStart`:
```typescript
private getPiecesOnUnusedLayersFromIndexToStart(startIndex: number, usedLayers: Set<string>, lifespans: PieceLifespan[]): Piece[] {
  return this.parts
    .slice(0, startIndex + 1)
    .flatMap(part => part.getPiecesWithLifespan(lifespans))
    .reduceRight(this.createGetPiecesOnUnusedLayersReducer(usedLayers), [])
}

private createGetPiecesOnUnusedLayersReducer(originalUsedLayers: Set<string>): (pieces: Piece[], piece: Piece) => Piece[] {
  const usedLayers = new Set(originalUsedLayers)
  return (pieces: Piece[], piece: Piece) => {
    if (!usedLayers.has(piece.layer)) {
      pieces.push(piece)
      usedLayers.add(piece.layer)
    }
    return pieces
  }
}
```

I have tested it and all tests (in sofie-server) passes :+1:

Reply:
Updated